---
title: "04: Practical: Core Approaches in _Bioconductor_"
author:
- name: Martin Morgan
  affiliation: Roswell Park Comprehensive Cancer Center
date: "`r format(Sys.time(), '%B %d, %Y')`"
vignette: >
  %\VignetteIndexEntry{04: Practical: Core Approaches in Bioconductor}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
  BiocStyle::html_document:
    number_sections: yes
    toc: true
---

```{r style, echo = FALSE, results = 'asis'}
knitr::opts_chunk$set(
    eval=as.logical(Sys.getenv("KNITR_EVAL", "TRUE")),
    cache=as.logical(Sys.getenv("KNITR_CACHE", "TRUE"))
)
options(width = 75)
```

# Answering useful questions with [GenomicRanges][]

```{r, message = FALSE}
library(GenomicRanges)
library(GenomicAlignments)
library(rtracklayer)
library(readr)
library(tibble)
```

The [IRanges][] package defines an important class for specifying
integer ranges, e.g.,
```{r iranges}
library(IRanges)
ir <- IRanges(start=c(10, 20, 30), width=5)
ir
```

There are many interesting operations to be performed on ranges, e.g,
`flank()` identifies adjacent ranges
```{r iranges-flank}
flank(ir, 3)
```

The `IRanges` class is part of a class hierarchy. To see this, ask R for
the class of `ir`, and for the class definition of the `IRanges` class
```{r iranges-class}
class(ir)
getClass(class(ir))
```

Notice that `IRanges` extends the `Ranges` class. Now try entering
`?flank` (`?"flank,<tab>"` if not using _RStudio_, where `<tab>` means
to press the tab key to ask for tab completion). You can see that
there are help pages for `flank` operating on several different
classes. Select the completion

```{r iranges-flank-method, eval=FALSE}
?"flank,Ranges-method"
```

and verify that you're at the page that describes the method relevant
to an `IRanges` instance.  Explore other range-based operations.

The [GenomicRanges][] package extends the notion of ranges to include
features relevant to application of ranges in sequence analysis,
particularly the ability to associate a range with a sequence name
(e.g., chromosome) and a strand. Create a `GRanges` instance based on
our `IRanges` instance, as follows
```{r granges}
library(GenomicRanges)
gr <- GRanges(c("chr1", "chr1", "chr2"), ir, strand=c("+", "-", "+"))
gr
```

The notion of flanking sequence has a more nuanced meaning in
biology. In particular we might expect that flanking sequence on the
`+` strand would precede the range, but on the minus strand would
follow it. Verify that `flank` applied to a `GRanges` object has this
behavior.
```{r granges-flank}
flank(gr, 3)
```

Discover what classes `GRanges` extends, find the help page
documenting the behavior of `flank` when applied to a `GRanges` object,
and verify that the help page documents the behavior we just observed.
```{r granges-class}
class(gr)
getClass(class(gr))
```
```{r granges-flank-method, eval=FALSE}
?"flank,GenomicRanges-method"
```
Notice that the available `flank()` methods have been augmented by the
methods defined in the _GenomicRanges_ package.

It seems like there might be a number of helpful methods available for
working with genomic ranges; we can discover some of these from the
command line.

```{r granges-methods, eval=FALSE}
methods(class="GRanges")
```

Use `help()` to list the help pages in the `GenomicRanges` package,
and `vignettes()` to view and access available vignettes; these are
also available in the RStudio 'Help' tab.

```{r granges-man-and-vignettes, eval=FALSE}
help(package="GenomicRanges")
vignette(package="GenomicRanges")
vignette(package="GenomicRanges", "GenomicRangesHOWTOs")
```

# Organizing data with [SummarizedExperiment][]

```{r, message = FALSE}
library(SummarizedExperiment)
library(DESeq2)
library(SingleCellExperiment)
```

## Working with an existing `SummarizedExperiment` object

The [airway][] experiment data package summarizes an RNA-seq
experiment investigating human smooth-muscle airway cell lines treated
with dexamethasone. Load the library and data set.

```{r}
library(airway)
data(airway)
airway
```

`airway` is an example of the _SummarizedExperiment_ class. Explore
its `assay()` (the matrix of counts of reads overlapping genomic
regions of interest in each sample), `colData()` (a description of
each sample), and `rowRanges()` (a description of each region of
interest; here each region is an ENSEMBL gene).

```{r}
x <- assay(airway)
class(x)
dim(x)
head(x)
colData(airway)
rowRanges(airway)
```

It's easy to subset a _SummarizedExperiment_ on rows, columns and
assays, e.g., retaining just those samples in the `trt` level of the
`dex` factor. Accessing elements of the column data is common, so
there is a short-cut.

```{r}
cidx <- colData(airway)$dex %in% "trt"
airway[, cidx]
## shortcut
airway[, airway$dex %in% "trt"]
```

It's also easy to perform range-based operations on
`SummarizedExperiment` objects, e.g., querying for range of chromosome
14 and then subsetting to contain only genes on this chromosome. Range
operations on rows are very common, so there are shortcuts here, too.

```{r}
chr14 <- as(seqinfo(rowRanges(airway)), "GRanges")["14"]
ridx <- rowRanges(airway) %over% chr14
airway[ridx,]
## shortcut
chr14 <- as(seqinfo(rowRanges(airway)), "GRanges")["14"]
airway[airway %over% chr14,]
```

The matrix of counts is available using the `assay()` accessor

```{r}
head(assay(airway))
```

1. Use the `assay()` and `rowSums()` function to remove all rows from
   the `airway` object that have 0 reads overlapping all
   samples. Summarize the library size (column sums of `assay()`) and
   plot a histogram of the distribution of reads per feature of
   interest.

2. Use `colSums()` to calculate the library size of each sample.

3. If `m` were a matrix, then `1 + m` would return another matrix with
   1 added to each element, and `log(m)` would return another matrix
   with log-transformed elements. Use these observations and
   `rowMeans()` to calculate a vector of average expression values of
   each gene. Visualize the result using `hist()`, `plot(density())`,
   or `ggplot() + geom_density()`.

## Constructing a `SummarizedExperiment` object 'by hand'

```{r, eval = FALSE}
pdata_file <- file.choos()    # airway-sample-sheet.csv
counts_file <- file.choose()  # airway-read-counts.csv
```
```{r, echo = FALSE}
pdata_file <-
    system.file(package="BiocIntro", "extdata", "airway-sample-sheet.csv")
counts_file <-
    system.file(package="BiocIntro", "extdata", "airway-read-counts.csv")
```

**Exercise** Use base _R_ `read.csv()` to import the sample sheet and
read counts. Use the argument `row.names=` to specific the `Run`
column to be used as row names

```{r}
pdata <- read_csv(pdata_file)
counts <- read_csv(counts_file)
```

**Exercise** Assemble `SummarizedExperiment` from transposed `counts`
and `pdata`. The "Run" column needs to be made into rownames.

```{r}
pdata <- column_to_rownames(pdata, "Run")
counts <- column_to_rownames(counts, "Run")
se <- SummarizedExperiment(t(counts), colData = pdata)
se
```

**Exercise** Following examples from the lectures, calculate total
library size and plot averge log gene expression from you
newly-created `SummarizedExperiment`.

# Integrative analysis with [MultiAssayExperiment][]

```{r, message = FALSE}
library(MultiAssayExperiment)
library(curatedTCGAData)
```

# Provenance

```{r}
sessionInfo()
```

[GenomicRanges]: https://bioconductor.org/packages/GenomicRanges
[SummarizedExperiment]: https://bioconductor.org/packages/SummarizedExperiment
[MultiAssayExperiment]: https://bioconductor.org/packages/MultiAssayExperiment
[airway]: https://bioconductor.org/packages/airway
[]: https://bioconductor.org/packages/
[]: https://bioconductor.org/packages/
