---
title: "Methylation Arrays"
author: Martin Morgan <martin.morgan@roswellpark.org>
date: "18 September, 2018"
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
vignette: >
  % \VignetteIndexEntry{Methylation Arrays}
  % \VignetteEngine{knitr::rmarkdown}
---

```{r style, echo = FALSE, results = 'asis'}
knitr::opts_chunk$set(
    eval=as.logical(Sys.getenv("KNITR_EVAL", "TRUE")),
    cache=as.logical(Sys.getenv("KNITR_CACHE", "TRUE"))
)
```

```{r setup, echo=FALSE}
suppressPackageStartupMessages({
    library(methylationArrayAnalysis)
    library(minfi)
    library(IlluminaHumanMethylation450kmanifest)
    library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
    library(limma)
})
```

# Introduction

- Based on array-based [Methylation Analysis Workflow][] 
- Public data from [GSE62944][].
- Data available in _R_ as part of the [methylationArrayAnalysis][] package

Methylation arrays

- Illumina 'beads' with particular 50bp DNA sequences, designed to interrogate specific genomic locations for presence of methylated CG dinucleotides.
- Methylated (M) and unmethylated (U) intensity, summarized as beta-values (M / (M + U), for visualization) or M-values (log2(M / U), for statistical analysis)

![](methylation-figure1.png)
<br />From Maksimovic, Gordon, and Oshlack, 2012


```{r}
datadir <- "~/extdata/methylation"
dir(datadir)
```

[Methylation Analysis Workflow]: https://bioconductor.org/packages/methylationArrayAnalysis
[GSE62944]: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE49667
[methylationArrayAnalysis]: https://bioconductor.org/packages/methylationArrayAnalysis

# Experimental design & upstream processing

- 10 samples. 
- 4 different sorted T-cell types (naive, rTreg, act\_naive, act\_rTreg)
- 3 different individuals (M28, M29, M30)

```{r}
sample_sheet <- dir(datadir, "SampleSheet.csv", full=TRUE)
samples <- read.csv(sample_sheet)
samples
```

# Data input

```{r}
library(minfi)
targets <- read.metharray.sheet(datadir, pattern="SampleSheet.csv")
methyl <- read.metharray.exp(targets=targets)
methyl
colData(methyl)[, c("Sample_Source", "Sample_Group")]
```

Recode two columns for later statistical analysis

```{r}
methyl$celltype <- factor(methyl$Sample_Group)
methyl$individual <- factor(methyl$Sample_Source)
```

# Quality control

```{r}
detP <- detectionP(methyl)
head(detP)
barplot(colMeans(detP))
```

# Normalization

```{r}
methyl_normalized <- preprocessQuantile(methyl)
```

```{r}
library(limma)     # plotMDS
plotMDS(getM(methyl_normalized))
```

# Filtering

FIXME: Remove probes that have failed in one or more suamples

FIXME: Remove sex-specific probes

FIXME: remove probes where common SNPs affect CpG.

FIXME: remove probes that are cross-reactive (i.e., mapping to multiple locations)

# Methylation analysis

## Probe-wise analysis

Model matrix

```{r}
design <- model.matrix(
    ~ 0 + celltype + individual,
    data=colData(methyl_normalized)
)
colnames(design) <- c(
    levels(methyl_normalized$celltype), 
    levels(methyl_normalized$individual)[-1]
)
```

fit

```{r}
fit <- lmFit(getM(methyl_normalized), design)
```

contrasts

```{r}
contrasts <- makeContrasts(
    naive - rTreg, naive - act_naive,
    rTreg - act_rTreg, act_naive - act_rTreg,
    levels=design
)
fit2 <- contrasts.fit(fit, contrasts)
```

Empirical Bayes

```{r}
fit2 <- eBayes(fit2)
```

Preview results

```{r}
summary(decideTests(fit2))
```

Find differentially methylated probes for first contrast

```{r}
dmprobes <- topTable(fit2, num=10, coef=1)
head(dmprobes)
```

# Comprehension

Probes to genes

```{r}
ann450k <- getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19)
```

```{r}
annotated <- merge(dmprobes, ann450k, by="row.names")
```

# Session info

As a part of the effort to provide reproducible analysis pipelines, it
is essential to document the software and versions used; here is the
information about _R_ packages used in the current chapter.

```{r}
sessionInfo()
```
