---
title: "A.3 -- Advanced _Bioconductor_"
author: Martin Morgan <martin.morgan@roswellpark.org>
date: "15 June, 2018"
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
vignette: >
  % \VignetteIndexEntry{A.3 -- Advanced Bioconductor}
  % \VignetteEngine{knitr::rmarkdown}
---

```{r style, echo = FALSE, results = 'asis'}
knitr::opts_chunk$set(
    eval=as.logical(Sys.getenv("KNITR_EVAL", "TRUE")),
    cache=as.logical(Sys.getenv("KNITR_CACHE", "TRUE"))
)
```
```{r setup, echo=FALSE}
suppressPackageStartupMessages({
    library(BiocParallel)
    library(HDF5Array)
    library(devtools)
    library(TENxBrainData)
    library(DelayedMatrixStats)
})
```

# S4 classes and methods

Basics

- A first class -- `setClass()`
- An essential method -- `setMethod()`

Commentary

- S3 versus S4
- Abstraction -- interface versus implementation

    - Constructors
    - Accessors

- Modelling columns (versus row)

Advanced

- Inheritance
- Multiple inheritance
- Function signatures

    - constant type
    - `...`

- Multiple dispatch

# Managing files -- BiocFileCache

Common use case

- files downloaded from the internet, stored 'somewhere'
- are they current? what if the internet changes? how can you share
  with others in your lab?
  
BiocFileCache

- maintains a local cache of files
- simplest use case -- add a resources

    ```{r}
    url <- "ftp://ftp.ensembl.org/pub/release-92/fasta/mus_musculus/cds/Mus_musculus.GRCm38.cds.all.fa.gz"
    local_file <- BiocFileCache::bfcrpath(rnames = url)
    ```
    
- first time: downloads to cache, returns local path. Subsequent:
  returns local path
- `BiocFileCache::bfcinfo()` for tibble of cache content
- Default cache location: `BiocFileCache::BiocFileCache()`, user home
  directory

Advanced use

- Cache could be anywhere, e.g., lab shared folder, project-specific folder.
- Cached resources could be the result of long-running calculations
- `BiocFileCache::bfcmeta()`: Arbitrary metadata can be associated with
  resources

# Managing memory -- DelayedArray

- R works on 'in memory' data, but... data can be too big to fit in memory.
- Traditional solutions, e.g., database, not always appropriate,
  especially for complex statistical modelling.
  
Example: 10x Genomics 'million neuron' single-cell RNA seq

- 27998 features measured on 1.3 million neuron cells (but actually
  only two biological individuals!)
  
One solution: on-disk representation that 'looks like' a matrix --
DelayedArray / HDF5Array.

- Available as an experiment-data package
- subsetting etc., fast and easy
- actual numerical computations performed block-wise, see
  DelayedMatrixStats package
  
```{r}
library(TENxBrainData)
library(DelayedMatrixStats)
tenx <- TENxBrainData()
tenx
tenx_1k <- tenx[, 1:1000]
hist(colSums(as.matrix(counts(tenx_1k))))
```

# Parallel computation -- BiocParallel

Priorities (ordered!)

1. Correct!
2. Tested -- reuse rather than reinvent!
3. Understandable
4. Performant
5. Parallel

BiocParallel

- Basic use
- Scaling through serial, core / computer, cluster computing models

```{r}
library(BiocParallel)

register(SerialParam())
bplapply(1:5, sqrt)

register(SnowParam(2))
bplapply(1:5, sqrt)

register(MulticoreParam(2)) # not on Windows
bplapply(1:5, sqrt)
```

```{r, eval = FALSE}
register(bpstart(BatchtoolsParam(2, cluster="sge")))
bplapply(1:5, sqrt)
```

Advanced BiocParallel: error recovery & debugging

```{r}
X <- list(1, 2, "three", 4)
register(MulticoreParam(2))
```

```{r, eval = FALSE}
fun <- function(x) sqrt(x)
res <- bptry(bplapply(X, fun))

debug(fun)
bplapply(X, fun, BPREDO = res, BPPARAM = SerialParam())
undebug(fun)

X[[3]] <- 3
bplapply(X, fun, BPREDO = res)
```


- Registering back ends
- As a package developer...
