---
title: "B.1 -- Annotation"
author: Martin Morgan <martin.morgan@roswellpark.org>
date: "16 June, 2018"
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
vignette: >
  % \VignetteIndexEntry{B.1 -- Annotation}
  % \VignetteEngine{knitr::rmarkdown}
---

```{r style, echo = FALSE, results = 'asis'}
knitr::opts_chunk$set(
    eval=as.logical(Sys.getenv("KNITR_EVAL", "TRUE")),
    cache=as.logical(Sys.getenv("KNITR_CACHE", "TRUE"))
)
```
```{r setup, echo=FALSE}
suppressPackageStartupMessages({
    library(org.Hs.eg.db)
    library(TxDb.Hsapiens.UCSC.hg18.knownGene)
    library(tidyverse)
    library(airway)
    library(DESeq2)
    library(biomaRt)
    library(KEGGREST)
    library(magrittr)
    library(VariantAnnotation)
})
```

# Introduction

Static resources

- org.*
- TxDb.*
- BSgenome.*
- Organism.dplyr

AnnotationHub and ExperimentHub
- ensembldb

Web resources

# `org.*`, `TxDb.*`, and `BSgenome.*` packages

Exercise 1: `org.*` packages

1. Install and attach the [org.Hs.eg.db][] annotation package; it
   contains 'symbol mapping' information for _Homo sapiens_, based on
   NCBI 'Entrez' identifiers.

    ```{r}
    library(org.Hs.eg.db)
    ```

2. Take a quick look at a summary of data in this package

    ```{r}
    org.Hs.eg.db
    ```
    
3. The idea is that there are `keytypes()` that can be mapped to
   different `columns()`; `keys()` can be used to see available
   keys. Explore the package to see what sorts of information is
   available, e.g.,
   
   ```{r}
   keytypes(org.Hs.eg.db)
   columns(org.Hs.eg.db)
   head(keys(org.Hs.eg.db, "SYMBOL"))
   ```
   
4. There are two basic ways of extracting data from an `org.*` package
   -- `mapIds()` to create a 1:1 mapping between key and a single
   column, and `select()` (it's often necessary to specify this
   function directly, to avoid a conflict with dplyr, as
   `AnnotationDbi::select()`). Explore these functions, e.g.,
   
   ```{r}
   set.seed(123)
   egid <- sample(keys(org.Hs.eg.db), 6)
   mapIds(org.Hs.eg.db, egid, "SYMBOL", "ENTREZID")
   select(org.Hs.eg.db, egid, c("SYMBOL", "ENSEMBL", "GENENAME"), "ENTREZID")
   ```
   
5. Some key - column mappings are 1:many, e.g., Entrez ID `"3812"`
   maps to 44 Ensembl Ids. What does `mapIds()` return when mapping
   Entrez ID `"3812"` to Ensembl ids? Use the additional argument
   `multiVals = "CharacterList"` to explore further. Compare results
   to those returned by `select()`.
   
   ```{r}
   egid <- "3812"
   mapIds(org.Hs.eg.db, egid, "ENSEMBL", "ENTREZID")
   mapIds(
       org.Hs.eg.db, egid, "ENSEMBL", "ENTREZID",
       multiVals = "CharacterList"
   )
   select(
       org.Hs.eg.db, egid, c("SYMBOL", "ENSEMBL"),
       multiVals = "CharacterList"
   )
   ```
   
6. It seems like it might often be useful to use the tidyverse on
   return values from `mapIds()` and `select()`; explore this usage
   
   ```{r}
   library(tidyverse)
   egid <- keys(org.Hs.eg.db)    # all ENTREZIDs
   mapIds(org.Hs.eg.db, egid, "SYMBOL", "ENTREZID") %>% 
      as_tibble() %>% 
      rownames_to_column("ENTREZID")
      
  AnnotationDbi::select(
      org.Hs.eg.db, egid, c("SYMBOL", "GO", "GENENAME"), "ENTREZID"
  ) %>% as_tibble()
  ```
  
[org.Hs.eg.db]: https://bioconductor.org/packages/org.Hs.eg.db

Exercise 2: `TxDb.*` packages

1. Install and attach the [TxDb.Hsapiens.UCSC.hg18.knownGene][]. This
   contains the gene models for _Homo sapiens_ based on the 'hg18'
   build of the human genome, using gene annotations in the UCSC
   'knownGene' annotation track; TxDb's for more recent builds and for
   different annotation tracks are available. Take a look at a summary
   of the package, and create an alias for easy typing
   
   ```{r}
   library(TxDb.Hsapiens.UCSC.hg18.knownGene)
   TxDb.Hsapiens.UCSC.hg18.knownGene
   txdb <- TxDb.Hsapiens.UCSC.hg18.knownGene
   ```
      
2. The main purpose of this package is to provide genomic coordinates
   of genomic features such as `exons()`, coding sequences (`cds()`),
   `transcripts()` and `genes()`. Explore, for example,
   
   ```{r}
   ex <- exons(txdb)
   ex
   library(ggplot2)
   qplot(log10(width(ex)))
   ex[ which.max(width(ex)) ]
   ```
   
3. Extract all genes, and then keep only the 'standard' chromosomes
   1:22, X, Y, and M. Use `table()` of `seqnames()` to determine how
   many genes are on each chromosome. Also do this in a dplyr way;
   note that the `seqnames(gn)` need to be coerced with `as.factor()`.
   
   ```{r}
   gn <- genes(txdb)
   length(gn)
   std <- paste0("chr", c(1:22, "X", "Y", "M"))
   seqlevels(gn, pruning.mode = "coarse") <- std
   length(gn)
   seqlevels(gn)
   
   table( seqnames(gn) )
   tibble(chr = as.factor(seqnames(gn))) %>% 
       group_by(chr) %>% 
       summarize(n = n())
   ```

4. `exonsBy()` groups exons by gene or transcript; extract exons
   grouped by gene. (Challenging!) can you identify genes with exons
   on different chromosomes? Are there any of these genes on the
   standard chromosomes?
   
   ```{r}
   exByGn <- exonsBy(txdb, "tx")
   ##
   trans <- lengths(unique(seqnames(exByGn)))
   table( trans )
   seqnames( exByGn[ trans > 1 ] )
   ##
   std <- paste0("chr", c(1:22, "X", "Y", "M"))
   unames <- unique(seqnames(exByGn[ trans > 1 ]))
   transstd <- all(unames %in% std)
   unames[transstd]
   ```

5. The previous exercise indicated that gene `"22947"` has exons on
   both chromosomes 4 and 10. Find out more about this gene using the
   [org.Hs.eg.db][] package and by searching for the gene symbol on
   the NCBI web site.

    ```{r}
    egid <- "22947"
    AnnotationDbi::select(
        org.Hs.eg.db, egid, c("SYMBOL", "GENENAME"), "ENTREZID"
    )
    ```
    
    ```{r, eval = FALSE}
    url <- paste0("https://www.ncbi.nlm.nih.gov/gene/", egid)
    browseURL(url)
    ```
    
6. Note that the `TxDb.*` packages also support `keytypes()`,
   `columns()`, and `select()` for mapping between exon, cds,
   transcript, and gene identifiers.
    
[TxDb.Hsapiens.UCSC.hg18.knownGene]: https://bioconductor.org/packages/TxDb.Hsapiens.UCSC.hg18.knownGene

Exercise 3: `BSgenome.*` packages

1. Install (if necessary) and load the [BSgenome.Hsapiens.UCSC.hg18][]
   package, containing the entire sequence of the hg18 build of _Homo
   sapiens_. Check out it's contents, and create a simple alias.
   
   ```{r}
   library(BSgenome.Hsapiens.UCSC.hg18)
   BSgenome.Hsapiens.UCSC.hg18
   hg18 <- BSgenome.Hsapiens.UCSC.hg18
   ```
   
2. Genomic sequence can be retrieved by chromosome, e.g.,
   `hg18[["chr1"]]`, or by genomic range, e.g., `getSeq(hg18,
   GRanges("chr1:1000000-2000000"))`. Retrieve your favorite chunk(s)
   of DNA and calculate GC content.
   
   ```{r}
   dna <- getSeq(hg18, GRanges("chr1:1000000-2000000"))
   letterFrequency(dna, "GC", as.prob=TRUE)
   ```

3. Use the `org.*`, `TxDb.*`, and `BSgenome.*` packages to retrieve
   the BRCA1 exon DNA sequence.
   
   ```{r}
   brca1_egid <- mapIds(org.Hs.eg.db, "BRCA1", "ENTREZID", "SYMBOL")
   brca1_exons <- exonsBy(txdb, "gene")[[brca1_egid]]
   getSeq(hg18, brca1_exons)
   ```

[BSgenome.Hsapiens.UCSC.hg18]: https://bioconductor.org/packages/BSgenome.Hsapiens.UCSC.hg18
   
# On-line resources, e.g., Biomart, KEGG

Exercise 1: [biomaRt][].

1. Explore the Biomart web site https://www.ensembl.org/biomart for
   retrieving all kinds of genomic annotations.
   
   Start by choosing a database (e.g., 'Ensembl Genes 92'), dataset
   (e.g., 'Human genes (GRCh38.p12)'), filter (e.g., 'GENE' / 'Input
   external reference' / 'Gene stable id' and enter
   'ENSG00000000003'), attributes (default is ok), then press
   'Results' to map from Ensembl identifier to transcript identifier.
   
2. Install (if necessary) and load the [biomaRt][] package. Use
   `listMarts()` to see availble databases, `useMart()` to select the
   mart you're interested in.

   ```{r}
   library(biomaRt)
   head(listMarts())
   mart <- useMart("ENSEMBL_MART_ENSEMBL")
   ```
   
3. Use `listDatasets()` and `useDataset()` to select the _Homo
   sapiens_ gene dataset.
   
   ```{r}
   head(listDatasets(mart))
   dataset <- useDataset("hsapiens_gene_ensembl", mart)
   ```
   
4. Use `listFilters()` to see available filters. The filter is the
   type of data that you are querying with. Choose one.

   ```{r}
   head(listFilters(dataset))
   filters <- "ensembl_gene_id"                    # see `listFilters()`
   ```

5. Use `listAttrbutes()` to see available attributes. Attributes
   represent the information you'd like to retrieve. Choose some!
   
   ```{r}
   head(listAttributes(dataset))
   attrs <- c("ensembl_gene_id", "hgnc_symbol")    # see `listAttributes()`
   ```

6. Create a character vector of Ensembl gene ids, compose and execute
   the query, transforming the result to a tibble.

    ```{r}
    ids <- c(
        "ENSG00000000003", "ENSG00000000005", "ENSG00000000419", 
        "ENSG00000000457", "ENSG00000000460", "ENSG00000000938"
    )
    tbl <- getBM(attrs, filters, ids, dataset) %>% as_tibble()
    tbl
    ```

[biomaRt]: https://bioconductor.org/packages/biomaRt

Exercise 2: [KEGGREST][]

1. Explore the KEGG web site https://www.genome.jp/kegg/ KEGG is a
   database of information on pathways.
   
2. Load the [KEGGREST][] package and discover available databases

    ```{r}
    library(KEGGREST)
    KEGGREST::listDatabases()
    ```
    
3. Use `keggList()` to query the pathway database for human pathways;
   present the result as a tibble
   
   ```{r}
   hsa_pathways <- keggList("pathway", "hsa") %>% 
       tibble(pathway = names(.), description = .)
   hsa_pathways
   ```
   
4. Use `keggLink()` to recover the genes in each pathway.

    ```{r}
    hsa_path_eg  <- keggLink("pathway", "hsa") %>% 
        tibble(pathway = ., egid = sub("hsa:", "", names(.)))
    hsa_path_eg
    hsa_path_eg %>% group_by(pathway) %>% summarize(genes = list(egid))
    ```

5. Update the `hsa_path_eg` table to include information on gene
   symbol and Ensembl id from the `org.Hs.eg.db` package. Retrieve the
   relevant information using `mapIds()`. How would you deal with
   entrez gene ids that map to multiple Ensembl ids?

    ```{r}
    hsa_kegg_anno <- hsa_path_eg %>%
        mutate(
            symbol = mapIds(org.Hs.eg.db, egid, "SYMBOL", "ENTREZID"),
            ensembl = mapIds(org.Hs.eg.db, egid, "ENSEMBL", "ENTREZID")
        )
    ```

6. Use `left_join()` to append pathway descriptions to the
   `hsa_kegg_anno` table.

    ```{r}
    left_join(hsa_kegg_anno, hsa_pathways)
    ```

[KEGGREST]: https://bioconductor.org/packages/KEGGREST

# [AnnotationHub][] and [ExperimentHub][]

Exercise 1: [AnnotationHub][]

1. The AnnotationHub contains a large number of annotation-style
   resources that supplement annotation packages like `TxDb.*`,
   etc. Install and attach the AnnotationHub package, and create a
   reference to the current hub; there are almost 45,000 resources in
   the hub.
   
   ```{r}
   library(AnnotationHub)
   hub <- AnnotationHub()
   hub
   ```

2. Use `query()` to search the hub for relevant resources. We'll
   search for `"EnsDb"` objects (these are `TxDb.*`-like databases of
   genome annotation data, based on Ensembl data) for `"Drosophila"`
   
   ```{r}
   query(hub, c("EnsDb", "Drosophila"))
   ```

3. Retrieve the most recent `EnsDb` for _Drosophila_; it might be
   necessary to install the [ensembldb][] package before the resource
   can be retrieved. The resource is stored in a local cache, so the
   data are only retrieved once.

    ```{r, message=FALSE}
    edb <- hub[["AH60963"]]
    ```

4. Figure out what to do with `edb` by using
   `browseVignettes(package="ensembldb")`, e.g.,
  
   ```{r}
   transcripts(edb, filter = ~ tx_biotype == "pseudogene")
   ```
  
   can you use `GRangesFilter()` to get only transcripts on 3L (hint:
   see `seqinfo(edb)` for chromosome widths).

5. Use `query()` to find other Ensembl-based resources for
   D. melanogaster, e.g.,
   
   ```{r}
   query(hub, c("ensembl", "92", "Drosophila"))
   ```

[AnnotationHub]: https://bioconductor.org/packages/AnnotationHub
[ensembldb]: https://bioconductor.org/packages/ensembldb

Exercise 2: [ExperimentHub][]

1. The ExperimentHub contains experimental data that can be
   incorporated into your own study, used in novel analysis, or used
   for training. They often have an associated _Bioconductor_ package
   that helps to 'manage' the resources. Load the [ExperimentHub][] package.
   
   ```{r}
   library(ExperimentHub)
   ```
   
   The [curatedTCGAData][] package is a _Bioconductor_ 'experiment
   data' package that helps to access TCGA resources on the
   ExperimentHub. Load this package
   
   ```{r}
   library(curatedTCGAData)
   ```
   
2. Discover TCGA resources by running the `curatedTCGAData()` function
   with the `dry.run = TRUE` argument.

    ```{r}
    curatedTCGAData(diseaseCode = "*", assays = "*", dry.run = TRUE)
    curatedTCGAData(diseaseCode = "COAD", assays = "RPPA*", dry.run = TRUE)
    ```

3. Retrieve TCGA data by specifying relevant disease code(s) and
   assay(s), e.g., all GBM (glioblastoma) samples assayed using
   [RPPA][] technology.

    ```{r, message = FALSE}
    gbm <- curatedTCGAData("GBM", "RPPA*", FALSE)
    ```

4. Explore the data -- it is a `MultiAssayExperiment` object, which in
   principle coordinates several different experiments conducted on
   overlapping samples. Here we have one experiment
   
   ```{r}
   experiments(gbm)[["GBM_RPPAArray-20160128"]]
   ```
   
   which is presented in the familiar `SummarizedExperiment` format.

[curatedTCGAData]: https://bioconductor.org/packages/curatedTCGAData
[ExperimentHub]: https://bioconductor.org/packages/ExperimentHub
[RPPA]: https://en.wikipedia.org/wiki/Reverse_phase_protein_lysate_microarray

# Annotating experimental results

Exercise 1: Generate a table of differentially expressed genes. 

1. Our main goal is to annotate experimental results, but we need to
   get some results to annotate! Load the [airway][] experiment data
   package and quickly run it through the DESeq pipeline
   
   ```{r, message = FALSE}
   library(DESeq2)
   library(airway)
   library(tidyverse)
   data(airway)
   airway$dex <- relevel(airway$dex, ref = "untrt")
   dds <- DESeqDataSet(airway, design = ~ cell + dex)
   dds <- DESeq(dds)
   ```
   
2. Extract the statisical results from the analysis, and coerce to a
   tibble. Remove rows that have `NA` adjusted p-values (these were
   rows that had no counts across all samples).
   
   ```{r}
   top <- results(dds) %>%
       as_tibble() %>% rownames_to_column("ensembl") %>%
       dplyr::filter(!is.na(padj))
   top
   ```
   
3. Create a 'volcano plot' of `-log10(padj)` ('statistically'
   differentially expressed) versus `log2FoldChange` ('biologically'
   differentially expressed)
   
   ```{r}
   library(ggplot2)
   ggplot(top, aes(log2FoldChange, -log10(padj))) + geom_point()
   ```

4. Use `mapIds()` from `org.Hs.eg.db` to add information about gene
   SYMBOL to the table.
   
   ```{r}
   top <- top %>% mutate(
       symbol = mapIds(org.Hs.eg.db, ensembl, "SYMBOL", "ENSEMBL"),
       entrez = mapIds(org.Hs.eg.db, ensembl, "ENTREZID", "ENSEMBL")
   )
   ```
   
5. Re-arrange the table to show the statistically most differentially
   expressed genes; show only the identifier columns, log2 fold
   change, and adjusted p-value.
   
   ```{r}
   top %>% arrange(padj) %>% 
      dplyr::select(ensembl, symbol, entrez, log2FoldChange, padj)
   ```
  
   Do the same for the biologically most differentiated.

   ```{r}
   top %>% arrange(desc(abs(log2FoldChange))) %>% 
       dplyr::select(ensembl, symbol, entrez, log2FoldChange, padj)
   ```
   
6. Update the volcano plot to annotate the five most statistically
   significantly differentiated genes with their gene symbol.

[airway]: https://bioconductor.org/packages/airway

# Gene set enrichment

INCOMPLETE WORK IN PROGRESS...

## What pathways?

GO

```{r}
ids <- top$ensembl
go <- AnnotationDbi::select(org.Hs.eg.db, ids, "GO", "ENSEMBL") %>%
    as_tibble()
## filter on specific ontology, evidence codes
```

KEGG

```{r}
kegg  <- keggLink("pathway", "hsa") %>% 
    tibble(pathway = ., entrez = sub("hsa:", "", names(.))) %>%
    mutate(
        ensembl = mapIds(
            org.Hs.eg.db, entrez, "ENSEMBL", "ENTREZID",
            multiVals = "list"
        )
    ) %>% unnest()
```

## Types of test

Self-contained -- evidence of consistent differential expression in this
particular pathway?

Competitive -- evidence of consistent differential expression in this
pathway, compared to expression in other pathways?

## Implementations

# Variant filtering

There is not enough time to cover this topic in detail; if you're
interested I suggest following the [VariantFiltering][] vignette.

[VariantFiltering]: https://bioconductor.org/packages/VariantFiltering
